FastText/src/models/custom_layers.jl

```diff
@@ -0,0 +1,410 @@
+    reset!(m)
+
+Resets the h, c parameters of the LSTM Cell.
+
+For more refer [`Flux.reset`](@ref https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.reset!)
+"""
+function reset!(m)
+    try		# to accomodate the definition in previously trained Language Model
+        (m.state = (m.cell.h, m.cell.c))
+    catch
+    	Flux.reset!(m)
+    end
+    Flux.reset!(m)
+end
+
+
+"""
+    reset_masks!(layer)
+
+This is an important funciton since it used to reset the masks
+which are saved in WeightDroppedLSTMCell after every pass.
+
+julia> wd = WeightDroppedLSTM()
+
+julia> reset_masks!(wd)
+"""
+function reset_masks!(wd::T) where T <: Flux.Recur{<:WeightDroppedLSTMCell}
+    wd.cell.maskWi = drop_mask(wd.cell.Wi, wd.cell.p)
+    wd.cell.maskWh = drop_mask(wd.cell.Wh, wd.cell.p)
+    return
+end
+####################################################################
+
+################## ASGD Weight-Dropped LSTM Layer ##################
+"""
+    AWD_LSTM(in::Integer, out::Integer, p::Float64=0.0)
+
+Average SGD Weight-Dropped LSTM
+
+This custom layer is used for training the Language model,
+instead of standard LSTM layer.
+This layer carries two addtional functionality:
+    Weight-dropping (DropConnect)
+    Averaging of weights
+
+AWD_LSTM is basically a wrapper aroung WeightDroppedLSTM layer,
+it has three fields:
+    layer : WeightDroppedLSTM layer
+    T     : Trigger iteration, to trigger averaging
+    accum : After triggring the accumlation of weights is saved here
+
+cite this paper to know more:
+https://arxiv.org/pdf/1708.02182.pdf
+"""
+mutable struct AWD_LSTM
+    layer::Flux.Recur
+    T::Integer
+    accum
+end
+
+AWD_LSTM(in::Integer, out::Integer, p::Float64=0.0; kw...) = AWD_LSTM(WeightDroppedLSTM(in, out, p; kw...), -1, [])
+
+Flux.@functor AWD_LSTM
+
+Flux.trainable(m::AWD_LSTM) = (m.layer,)
+
+(m::AWD_LSTM)(in) = m.layer(in)
+
+# To set the trigger point for the AWD_LSTM layer
+set_trigger!(t, m) = nothing
+set_trigger!(trigger_point::Integer, m::AWD_LSTM) = m.T = trigger_point;
+
+# resets mask of the WeightDroppedLSTM layer contained in AWD_LSTM
+reset_masks!(awd::AWD_LSTM) = reset_masks!(awd.layer)
+
+"""
+    asgd_step!(i::Integer, layer::AWD_LSTM)
+
+Averaged Stochastic Gradient Descent Step
+
+This funciton performs the Averaging step to the given AWD_LSTM layer,
+if the trigger point or trigger iteration is reached.
+Arguments:
+    i       : current iteration of the training loop
+    layer   : AWD_LSTM layer
+"""
+asgd_step!(i, l) = nothing
+
+function asgd_step!(iter::Integer, layer::AWD_LSTM)
+    if (iter >= layer.T) & (layer.T > 0)
+        p = get_trainable_params([layer])
+        avg_fact = 1/max(iter - layer.T + 1, 1)
+        if avg_fact != 1
+            layer.accum = layer.accum .+ p
+            for (ps, accum) in zip(p, layer.accum)
+                ps .= avg_fact*accum
+            end
+        else
+            layer.accum = deepcopy(p)   # Accumulator for ASGD
+        end
+    end
+    return
+end
```

Comment:
@lorenzoh I don't believe we have any utils for this in Flux or FastAI, right? Haven't kept up with the list of callbacks.